kvdb/postgres/readwrite_tx.go

```diff
@@ -0,0 +1,386 @@
+//
+// Furthermore there is a <table>_p index on parent_id that is required for the
+// foreign key constraint.
+//
+// Finally there are unique indices on (parent_id, key) to prevent the same key
+// being present in a bucket more than once (<table>_up and <table>_unp). In
+// postgres, a single index wouldn't enforce the unique constraint on rows with
+// a NULL parent_id. Therefore two indices are defined.
+func getCreateTableSql(table string) string {
+	return `
+CREATE TABLE IF NOT EXISTS public.` + table + `
+(
+    key bytea NOT NULL,
+    value bytea,
+    parent_id bigint,
+    id bigserial PRIMARY KEY,
+    sequence bigint,
+    CONSTRAINT ` + table + `_parent FOREIGN KEY (parent_id)
+        REFERENCES public.` + table + ` (id)
+        ON UPDATE NO ACTION
+        ON DELETE CASCADE
+);
+
+CREATE INDEX IF NOT EXISTS ` + table + `_p
+    ON public.` + table + ` (parent_id);
+
+CREATE UNIQUE INDEX IF NOT EXISTS ` + table + `_up
+    ON public.` + table + `
+    (parent_id, key) WHERE parent_id IS NOT NULL;
+
+CREATE UNIQUE INDEX IF NOT EXISTS ` + table + `_unp 
+    ON public.` + table + ` (key) WHERE parent_id IS NULL;
+`
+}
+
+// CreateTopLevelBucket creates the top level bucket for a key if it
+// does not exist.  The newly-created bucket it returned.
+func (tx *readWriteTx) CreateTopLevelBucket(key []byte) (walletdb.ReadWriteBucket, error) {
+	if len(key) == 0 {
+		return nil, walletdb.ErrBucketNameRequired
+	}
+
+	table := tx.toTableName(key)
+
+	// If the key can't be mapped to a table name, open the bucket from the
+	// hex table.
+	if table == "" {
+		bucket := newReadWriteBucket(tx, tx.db.hexTableName, nil)
+		return bucket.CreateBucketIfNotExists(key)
+	}
+
+	createTableSql := getCreateTableSql(table)
+	_, err := tx.Exec(createTableSql)
+	if err != nil {
+		return nil, err
+	}
+
+	return newReadWriteBucket(tx, table, nil), nil
+}
+
+// DeleteTopLevelBucket deletes the top level bucket for a key.  This
+// errors if the bucket can not be found or the key keys a single value
+// instead of a bucket.
+func (tx *readWriteTx) DeleteTopLevelBucket(key []byte) error {
+	table := tx.toTableName(key)
+
+	if table != "" {
+		// This top-level bucket should be stored in a distinct table.
+		// Verify that this table exists.
+		var dummy int
+		err := tx.QueryRow(
+			"SELECT 1 FROM pg_catalog.pg_tables "+
+				"WHERE schemaname='public' AND tablename=$1",
+			table,
+		).Scan(&dummy)
+
+		switch {
+		case err == sql.ErrNoRows:
+			return walletdb.ErrBucketNotFound
+
+		case err != nil:
+			return err
+		}
+
+		// Drop the table.
+		_, err = tx.Exec(
+			"DROP TABLE public." + table + ";",
+		)
+		if err != nil {
+			return err
+		}
+
+		// Remove the table sequence number.
+		_, err = tx.Exec(
+			"DELETE FROM "+tx.db.sequenceTableName+
+				" WHERE table_name=$1", table,
+		)
+		return err
+	}
+
+	// This top-level bucket is located in the hex table. Execute a
+	// cascading delete on the key.
+	result, err := tx.Exec(
```

Comment:
Just to reassure, this will delete the sequence too?

Reply:
Yes, the sequence is part of that same row in the hex table. Only for human-readable tl bucket keys, we use the separate sequence table.